(Design) Formatted Parts (#463)
unicode-org/message-format-wg: exploration/formatted-parts.md

  at least as far as that's possible and appropriate.

## Proposed Design

The formatted-parts API is included in the spec as an optional but recommended formatting target.

The shape of the formatted-parts output is defined in a manner similar to the data model,
which includes TypeScript, JSON Schema, and XML DTD definitions of the same data structure.

At the top level, the formatted-parts result is an iterable sequence of parts.
Parts corresponding to each _text_ can be simpler than those of _expressions_,
as they do not have a `source` other than their `value`,
or set any of the other possible metadata fields.

```ts
type MessageParts = Iterable<
  MessageTextPart | MessageExpressionPart | MessageBiDiIsolationPart
>;

interface MessageTextPart {
  type: "text";
  value: string;
}
```

For MessageExpressionPart, the `source` corresponds to the expression's fallback value.
The `dir` and `locale` attributes of a part may be inherited from the message
or from the operand (if present),
or overridden by an expression attribute or formatting function,
or otherwise set by the implementation.
Each part should have at most one of `value` or `parts` defined;

> some may have none.

Reviewer: While this may be how this gets implemented in certain implementations, I think our definition should be stricter. Something like:

```ts
type MessageExpressionPart = MessageSingleValuePart | MessageMultiValuePart;

interface MessageSingleValuePart {
  type: string;
  source: string;
  value?: unknown;
  dir?: "ltr" | "rtl" | "auto";
  locale?: string;
}

interface MessageMultiValuePart {
  type: string;
  source: string;
  parts?: Iterable<{ type: string; value: unknown; source?: string }>;
  dir?: "ltr" | "rtl" | "auto";
  locale?: string;
}
```

Author: This sounds like a sensible way to specify this, should this design doc be accepted. Thought with a required `parts` to explicitly differentiate the two.

Author: Added the interfaces.

```ts
interface MessageExpressionPart {
  type: string;
  source: string;

>   parts?: Iterable<{ type: string; value: unknown }>;

Author: ```suggestion
  parts?: Iterable<{ type: string; value: unknown; source?: string }>;
```

>   parts?: Iterable<{ type: string; value: unknown; source?: string }>;

Reviewer: Let's call this `subparts`, `items`, `values`, `chunks`, `fragments`... anything but `parts`, which we already use to describe the things returned by the `formatToParts` API.

>   value?: unknown;

Reviewer: I think it would be more robust to describe the polymorphism of parts through generics rather than `unknown`. Something like the following:

```ts
interface MessageSingleValuePart<T> {
  type: string;
  source: string;
  value: T;
  dir?: "ltr" | "rtl" | "auto";
  locale?: string;
}

type MessageStringPart = MessageSingleValuePart<string>;
```

Reviewer: I mean to use it as a language for describing the concepts introduced by this doc; `MessageStringPart` is one example.

Author: I added generic interfaces in line with your suggestion, but they're now referred to in the doc via comments, so that each part can retain its full definition in-place and a reader doesn't need to (but may) look up how they relate to each other. This also allows us to continue expressing how e.g. MessageFallbackPart explicitly does not have `locale` and `dir` fields, even if it does otherwise extend `MessageSingleValuePart<"fallback", never>`.

```ts
interface MessageBiDiIsolationPart {
  type: "bidiIsolation";
  value: "\u2066" | "\u2067" | "\u2068" | "\u2069"; // LRI | RLI | FSI | PDI
}
```

Some of the MessageExpressionPart instances may be further defined
without reference to the function registry.

Unannotated expressions with a _literal_ operand
are represented by MessageStringPart.
As with MessageTextPart,
the `value` of MessageStringPart is always a string.

```ts
interface MessageStringPart {
  type: "string";
  source: string;
  value: string;
  dir?: "ltr" | "rtl" | "auto";
  locale?: string;
}
```

Unannotated expressions with a _variable_ operand
whose type is not recognized by the implementation
or for which no default formatter is available
are represented by MessageUnknownPart.

```ts
interface MessageUnknownPart {
  type: "unknown";
  source: string;
  value: unknown;
}
```

When the resolution or formatting of a placeholder fails,
it is represented in the output by MessageFallbackPart.
No `value` is provided; when formatting to a string,
the part's representation would be `'{' + source + '}'`.

```ts
interface MessageFallbackPart {
  type: "fallback";
  source: string;
}
```

> Formatting functions defined in the registry

Reviewer: ```suggestion
### Formatting functions defined in the registry
```

Reviewer: Or, perhaps even better:

```suggestion
### Registry definition of formatted parts
```

Each function defined in the registry MUST define its "formatted-parts" representation.
A function can define either a unitary string `value` or a `parts` representation.
Where possible, a function SHOULD provide a `parts` representation
if its output might reasonably consist of multiple fields.
Where available, such a formatted value should itself be represented by `parts`

> rather than a unitary string `value`.

Reviewer: What does this mean, given the sentence above it? The design doesn't allow both `value` and `parts` to be defined on a single part, does it?

Author: I'm pretty sure this is a remnant of some editorial work, and it's saying the same as the previous sentence but more complicatedly.

```suggestion
```

> These sub-parts should not need fields beyond their `type` and `value`,
> and in most cases it's presumed that the sub-part `value` would be a string.

Author: ```suggestion
In most cases, these sub-parts should not need fields beyond their `type` and a string `value`,
Where necessary, other `value` types may be used and other fields such as a `source` included.
```

  // MessageSingleValuePart<"string", string>
  type: "string";
  source: string;
  value: string;
  dir?: "ltr" | "rtl" | "auto";
  locale?: string;
}
```

Unannotated expressions with a _variable_ operand
whose type is not recognized by the implementation
or for which no default formatter is available
are represented by MessageUnknownPart.

```ts
interface MessageUnknownPart {
  // MessageSingleValuePart<"unknown", unknown>
  type: "unknown";
  source: string;
  value: unknown;
}
```

When the resolution or formatting of a placeholder fails,
it is represented in the output by MessageFallbackPart.
No `value` is provided; when formatting to a string,
the part's representation would be `'{' + source + '}'`.

```ts
interface MessageFallbackPart {
  // MessageSingleValuePart<"fallback", never>
  type: "fallback";
  source: string;
}
```

### Registry definition of formatted parts

Each function defined in the registry MUST define its "formatted-parts" representation.
A function can define either a unitary string `value` or a `parts` representation.
Where possible, a function SHOULD provide a `parts` representation
if its output might reasonably consist of multiple fields.

> In most cases, these sub-parts should not need fields beyond their `type` and a string `value`,
> Where necessary, other `value` types may be used and other fields such as a `source` included.

Author: ```suggestion
In most cases, these sub-parts should not need fields beyond their `type` and a string `value`.
Where necessary, other `value` types may be used
and other fields such as a `source` included in the sub-parts,
and additional fields included in the `MessageExpressionPart`.
```

whose type is not recognized by the implementation
or for which no default formatter is available
are represented by MessageUnknownPart.

```ts
interface MessageUnknownPart {
  // MessageSingleValuePart<"unknown", unknown>
  type: "unknown";
  source: string;
  value: unknown;
}
```

When the resolution or formatting of a placeholder fails,
it is represented in the output by MessageFallbackPart.
No `value` is provided; when formatting to a string,
the part's representation would be `'{' + source + '}'`.

```ts
interface MessageFallbackPart {
  // MessageSingleValuePart<"fallback", never>
  type: "fallback";
  source: string;
}
```

### Registry definition of formatted parts

Each function defined in the registry MUST define its "formatted-parts" representation.
A function can define either a unitary string `value` or a `parts` representation.
Where possible, a function SHOULD provide a `parts` representation
if its output might reasonably consist of multiple fields.
In most cases, these sub-parts should not need fields beyond their `type` and a string `value`.
Where necessary, other `value` types may be used
and other fields such as a `source` included in the sub-parts,

> and additional fields included in the `MessageExpressionPart`.

Author: ```suggestion
and additional fields may be included in the `MessageExpressionPart`.
```
